### PR TITLE
Registration,Login 

### PR DESCRIPTION
**Pull request by GigiDev7**

### CODE REVIEW

File: server/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+const handleError = (error) => {
+  let errors = {
+    email: error?.errors?.email?.message || "",
+    password: error?.errors?.password?.message || "",
+    firstname: error?.errors?.firstname?.message || "",
+    lastname: error?.errors?.lastname?.message || "",
+    dateOfBirth: error?.errors?.dateOfBirth?.message || "",
+  };
+  if (error.code === 11000) {
+    errors.email = "Email already exists";
```
Comment (Lokki17) [inline server/controllers/user.js]: Why do you set text 'Email already exists' as email?

File: server/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  return errors;
+};
+
+const login = async (req, res) => {
+  try {
```
Comment (Lokki17) [inline server/controllers/user.js:9]: Move logic from controller to service layer

File: server/controllers/user.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    });
+  } catch (err) {
+    res.status(500).json(err);
+  }
+};
+
+const register = async (req, res) => {
+  try {
+    await createUser(req.body);
+    res.status(200).json({ message: "Successfully registered" });
```
Comment (Lokki17) [inline server/controllers/user.js]: Please return created user instead of text in body

File: server/app.js [hunk 4]
```diff
@@ -2,6 +2,13 @@ const express = require("express");
...
 require("dotenv").config();
 
+const userRouter = require("./routes/user");
+
 const app = express();
 
+app.use(express.json());
+app.use(cors());
+
+app.use("/api/user", userRouter);
```
Comment (Lokki17) [inline server/app.js]: remove 'api'

File: server/middlewares/authGuard.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  if (!req.headers.authorization || !req.headers.authorization.split("")[1]) {
+    return res.status(403).json({ message: "Unathorized" });
+  }
+
+  const token = req.headers.authorization.split("")[1];
+  jwt.verify(token, process.env.JWT_SECRET, (err, decodedData) => {
+    if (err) {
+      return res.status(401).json({ message: "Invalid token" });
+    }
+    req.userId = decodedData.id;
```
Comment (Lokki17) [inline server/middlewares/authGuard.js]: It's better to add 'user' in request. User can has role, firstName, lastName and so on

File: server/services/user.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  return token;
+};
+
+const comparePasswords = async (password, hashedPassword) => {
+  const isPasswordCorrect = await bcrypt.compare(password, hashedPassword);
+  return isPasswordCorrect;
+};
+
+const findUser = async (email) => {
+  const user = await User.findOne({ email });
```
Comment (Lokki17) [inline server/services/user.js:20]: return User.findOne({ email });

File: server/services/user.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  return user;
+};
+
+const createUser = async (userData) => {
+  await User.create(userData);
```
Comment (Lokki17) [inline server/services/user.js]: Why dont you return created user?

File: server/services/user.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const findAndDeleteUser = async (userId) => {
+  await User.findByIdAndDelete(userId);
```
Comment (Lokki17) [inline server/services/user.js:41]: return something

File: server/services/user.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const findUserAndUpdate = async (userId, userData) => {
+  await User.findByIdAndUpdate(userId, userData);
```
Comment (Lokki17) [inline server/services/user.js]: return updated user

File: server/middlewares/adminGuard.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+const User = require("../models/userSchema");
+
+const adminGuard = (req, res, next) => {
+  const user = User.findById(req.userId);
```
Comment (Lokki17) [inline server/middlewares/adminGuard.js]: If you add from token user in request you will need to fetch user from DB

**Review by Lokki17 — CHANGES_REQUESTED** [review]
Fix remarks

File: server/controllers/user.js (hunk 1 above)
Comment (GigiDev7) [reply server/controllers/user.js:19]: what should it be?

Comment (GigiDev7) [reply server/controllers/user.js]: error is related to email field

Comment (Lokki17) [reply server/controllers/user.js]: To be honest I dont like this. It's not validation issue. I'd like to throw exception with code and handle it in common exception handler

Comment (GigiDev7) [reply server/controllers/user.js]: can you show me an example for that?

**Pull request merged**